codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ConstrainedCollectionGenerator.kt

```diff
@@ -180,14 +179,145 @@ class ConstrainedCollectionGenerator(
+                    rust(
+                        """
+                        Self::UniqueItems { duplicate_indices, .. } => 
+                            crate::model::ValidationExceptionField {
+                                message: format!("${uniqueItemsTrait.validationErrorMessage()}", &duplicate_indices, &path),
+                                path,
+                            },
+                        """,
+                    )
+                },
+                validationFunctionDefinition = { constraintViolation, _ ->
+                    {
+                        // This is the fun bit where we enforce the trait.
+                        //
+                        // The algorithm to check for uniqueness is fairly simple: we iterate over the incoming items
+                        // in order and store them in a `HashMap`, with an entry whose key is the item and whose value
+                        // is the most recently seen index where we encountered the item. If we encounter an item we've
+                        // seen before, we add its index to a `duplicate_indices` vector. Lastly, we iterate over the
+                        // `duplicate_indices` and lookup each of them in the seen items hash map, to extend the
+                        // duplicates with those duplicate items that were seen last.
+                        //
+                        // The algorithm has linear time complexity over the list's length. In the worst case,
+                        // when all items have the same value, we'd iterate over all of the items twice and use
+                        // linear memory.
+                        //
+                        // We can use a `HashMap` to store the incoming items because it is guaranteed that all
+                        // shapes in the list shape's closure implement `Hash`; see
+                        // [`DeriveEqAndHashSymbolMetadataProvider`]. Indeed, smithy-rs only allows the `@uniqueItems`
+                        // trait to be applied to a list shape if and only if all the shapes in the list shape's closure
+                        // implement `Eq` and `Hash`.
+                        rustTemplate(
+                            """
+                            fn check_unique_items(items: #{Vec}<#{MemberSymbol}>) -> Result<#{Vec}<#{MemberSymbol}>, #{ConstraintViolation}> {
+                                let mut seen = #{HashMap}::new();
+                                let mut duplicate_indices = #{Vec}::new();
+                                for (idx, item) in items.iter().enumerate() {
+                                    if let Some(prev_idx) = seen.insert(item, idx) {
+                                        duplicate_indices.push(prev_idx);
+                                    }
+                                }
+
+                                let mut last_duplicate_indices = #{Vec}::new();
```

Comment:
Is there a way to convince the compiler that I'll be pushing to disjoint slices of `duplicate_indices` in order to avoid allocating this vector here? 🤔

Reply:
Do you want to modify `duplicate_indices` in place, where you give the assurance that you are going to push after the original `len()` of `duplicate_indices`? You could probably do it with unsafe if we uphold the requirements above. Is there a way in safe rust? 🤔